src/services/polkit/qml.cpp

```diff
@@ -0,0 +1,428 @@
+
+	if (mInputRequest) {
+		mInputRequest->deleteLater();
+	}
+
+	mInputRequest = new InputRequest(message, echo, currentSession);
+	emit inputRequestChanged();
+}
+
+void PolkitAgent::completed(bool gainedAuthorization) {
+	qCDebug(logPolkit) << "authentication request completed";
+
+	auto& req = *queuedRequests.front();
+	if (gainedAuthorization) {
+		emit authenticationSucceeded();
+
+		req.complete();
+		finishAuthenticationRequest();
+	} else if (isCancelled) {
+		req.cancel("Authentication request was cancelled");
+		finishAuthenticationRequest();
+	} else {
+		emit authenticationFailed();
+
+		setupSession();
+	}
+}
+
+void PolkitAgent::showError(const QString& message) {
+	qCDebug(logPolkit) << "showing error message:" << message;
+
+	if (mSubMessage) {
+		mSubMessage->deleteLater();
+	}
+
+	mSubMessage = new SubMessage(message, true, currentSession);
+	emit subMessageChanged();
+}
+
+void PolkitAgent::showInfo(const QString& message) {
+	qCDebug(logPolkit) << "showing info message:" << message;
+
+	if (mSubMessage) {
+		mSubMessage->deleteLater();
+	}
+
+	mSubMessage = new SubMessage(message, false, currentSession);
+	emit subMessageChanged();
+}
+
+void PolkitAgent::activateAuthenticationRequest() {
+	if (queuedRequests.empty()) {
+		return;
+	}
+
+	AuthRequest& req = *queuedRequests.front();
+
+	qCDebug(logPolkit) << "activating authentication request for action" << req.actionId;
+
+	for (auto identity: mIdentities) {
+		delete identity;
+	}
+	mIdentities.clear();
+
+	for (auto identity: req.identities) {
+		Identity* obj;
+
+		if (POLKIT_IS_UNIX_USER(identity)) {
+			auto uid = polkit_unix_user_get_uid(POLKIT_UNIX_USER(identity));
+			auto pw = getpwuid(uid);
+			auto name = (pw && pw->pw_name && *pw->pw_name) ? QString::fromUtf8(pw->pw_name)
+			                                                : QString::number(uid);
+			QString icon;
+			if (pw && pw->pw_dir && *pw->pw_dir) {
+				icon = QString::fromUtf8(pw->pw_dir) + QDir::separator() + ".face.icon";
+				if (!QFile::exists(icon)) {
+					icon.clear();
+				}
+			}
+			obj = new Identity(
+			    uid,
+			    name,
+			    (pw && pw->pw_gecos && *pw->pw_gecos) ? QString::fromUtf8(pw->pw_gecos) : name,
+			    icon,
+			    false,
+			    identity
+			);
+		}
+
+		if (POLKIT_IS_UNIX_GROUP(identity)) {
+			auto gid = polkit_unix_group_get_gid(POLKIT_UNIX_GROUP(identity));
+			auto gr = getgrgid(gid);
+			auto name = (gr && gr->gr_name && *gr->gr_name) ? QString::fromUtf8(gr->gr_name)
+			                                                : QString::number(gid);
+			obj = new Identity(
+			    gid,
+			    name,
+			    name,
+			    QString(), // no icon for groups
+			    true,
+			    identity
+			);
+		}
+
+		// For now, other identity types (currently only PolkitUnixNetgroup)
+		// are not supported.
+
+		if (obj) {
+			mIdentities.append(obj);
+		}
+	}
+
+	mSelectedIdentity = mIdentities.isEmpty() ? nullptr : mIdentities.first();
```

Comment:
I'm not sure how much significance it has with polkit, but would it make sense for a user to pick any other identity if multiple are available?

Reply:
Then we should probably expose that as an option

Reply:
The user can set `selectedIdentity` at any time, even after authentication is already started. This is just a (in my opinion) reasonable default. The alternative would be to not set this and always make the user choose.

This behavior also mirrors what all existing PolKit agents that I'm aware of do.